Volume Provisioning E2E: test PVC delete causes PV delete

copejon opened the pull request:
**What this PR does / why we need it**:
Test for a regression addressed in #21268.  There was a case where the PVC being created and deleted quickly may result in a provisioned PV left behind as `Available.`

```release-note
NONE
```
 
cc @jeffvance

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41156)
<!-- Reviewable:end -->

krousey commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +125,41 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+
+			// The provisioner uses a deterministic convention of "pvc-" + the claim UID for volume names,
+			// so we do not need to wait for the claim to bind to know the volume.
+			pv := "pvc-" + string(claim.UID)
```
If this is not a guarantee, please do not rely on it in tests.

krousey commented:
cc @kubernetes/sig-storage-pr-reviews Can someone give a storage review?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +125,41 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			// Wait for the claim to create before deleting it.
+			err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimPending, c, ns, claim.Name, 1*time.Second, 60*time.Second)
```
won't the claim bind to the dyn provisioned pv, so the claim's phase will be "Bound"?

copejon replied:
The pvc while sit in Pending until the volume is provisioned, then be Bound by the controller.  The idea here was just to wait until the claim is pending before deleting it, but I don't think it's actually necessary at all.  Will remove it.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +125,41 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			// The provisioner uses a deterministic convention of "pvc-" + the claim UID for volume names,
+			// so we do not need to wait for the claim to bind to know the volume.
```
Can't the wait above can potentially miss the "Released" phase?  IOW, the pvc goes from "Released" -> "Bound" during the wait interval (I think).

copejon replied:
see above comment.  The claim will create as Pending until it is Bound.  We don't want to let the PVC bind though otherwise we'll have missed the window for the race.

jeffvance replied:
Understood. But I'm saying that I _think_ using polling here you can miss the pvc's phase going from Pending -> Bound. You can miss "Pending" and just see "Bound", depending on timing.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +125,41 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			// so we do not need to wait for the claim to bind to know the volume.
+			pv := "pvc-" + string(claim.UID)
+			// In case of a regression, manually delete the volume
+			defer c.Core().PersistentVolumes().Delete(pv, nil)
```
No `Except` here?

jsafrane commented:
@k8s-bot ok to test

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +125,38 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 			testDynamicProvisioning(c, claim)
 		})
+
+		It("should test that deleting a claim before the volume is provisioned deletes the volume. [Slow][Volume]", func() {
```
Why is it [slow]? I don't see anything particularly slow (but see below...)

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +125,38 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+
+		It("should test that deleting a claim before the volume is provisioned deletes the volume. [Slow][Volume]", func() {
+			// This case tests for the regressions of a bug fixed by PR #21268
+			// REGRESSION: Deleting the PVC before the PV is provisioned can result in PV
```
This was a race condition, i.e. perfect timing of PVC deletion in the middle of provisioning was required. By creating just one PVC, the probability of catching this is really low. What about creating 10 - 20 (- 100?) PVCs?

copejon replied:
Agreed.  Jeff and I actually hit on this yesterday also.  I've made some changes to iteratively create 100 PVCs then List PVs and check for any matching the pseudorandom storage class name.  Will push today after some testing / tweaking.

jsafrane reviewed (commented):
The code looks solid, on the other hand we could increase probability of catching the race condition it is supposed to test.

jsafrane commented:
@k8s-bot verify test this

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+
+			By(fmt.Sprintf("Creating and deleting PersistentVolumeClaims for %d iterations", raceAttempts))
+			for i := 0; i < raceAttempts; i++ {
+				claim := newClaim(ns, false)
```
do you want a new claim obj for each iteration, or to reuse the same obj? Former has more GC work...

copejon replied:
Good point.  I was thinking it would result in name collisions but that doesn't get assigned until the Create call.  Will tweak that.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			}
+
+			By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
+			clusterPVs, err := c.Core().PersistentVolumes().List(metav1.ListOptions{})
```
Is there any filtering avail in `ListOptions`?

copejon replied:
There is a `FieldSelector` option.  I haven't been able to target the storage class annotation successfully.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				if storageutil.HasStorageClassAnnotation(pv.ObjectMeta) &&
+					pv.Annotations[storageutil.StorageClassAnnotation] == class.Name &&
+					pv.Status.Phase != v1.VolumeReleased {
+					// Any PVs associated with this storage class that are not Released
```
Does "Released" indicate that the pv will be deleted shortly?

copejon replied:
The controller has a process that checks for the Released phase executes the appropriate reclaim policy in [pv_controller.go](https://github.com/copejon/kubernetes/blob/923a3d59b3dc39b701784f13f12ba7a62411e579/pkg/controller/volume/persistentvolume/pv_controller.go#L1064).

jsafrane replied:
Do not check the status at all, just in case there is a bug in the controller. Just presence of the PV is enough to see that something is wrong.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			}
+			// Cleanup associated pv and disk
+			defer func() {
+				for _, pv := range regressedPVs {
```
so "Released" pvs are skipped here

copejon replied:
PVs marked `Released` must have been marked as a result of the PVC delete call.  Ones marked `Available` are an indicator of regression.  Now if they're `Bound`, then it could be a latency issue.  I may need to add a wait in there.

jeffvance replied:
`or _, pvName := range regressedPVs {...`

jsafrane replied:
>  I may need to add a wait in there.

The test already waits for 5 minutes, that should be enough to delete everything.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			class := newStorageClass()
+			class.Name = ns + "-sc" // pseudo-random Storage Class name
+			class, err := c.Storage().StorageClasses().Create(class)
+			defer c.Storage().StorageClasses().Delete(class.Name, nil)
```
I think the `defer` should be after the `Expect`. We don't want to delete a SC that isn't created.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			claim := newClaim(ns, false)
+			claim.Annotations[storageutil.StorageClassAnnotation] = class.Name
+			for i := 0; i < raceAttempts; i++ {
+				tmpClaim, err := c.Core().PersistentVolumeClaims(ns).Create(claim)
```
Same comment except that var name is "tmpClaim" rather than "claim". The point I'm trying to make (though I might be wrong) is that you are creating a new object for every iteration in the `for`. Now, it's not a big deal because that object is just a ptr in this case. I was thinking it would be "better" to define `claim` outside the loop and just assign it using `=` in the loop. However, using `:=` in `for` loops is common in k8s code, so that's all I'll say on this.

copejon replied:
The issue is that claims will be created with a random name on the `.Create()` call which we need to capture for the `.Delete()` call.  Also, if there is any latency in the delete call and the PVCs are using a predefined name, it's possible to cause a creating PVC to collide with a PVC waiting for deletion.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				}
+			}
+			// Cleanup associated pv and disk
+			defer func() {
```
nice!

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
+			clusterPVs, err := c.Core().PersistentVolumes().List(metav1.ListOptions{})
+			regressedPVs := make([]v1.PersistentVolume, 0)
+			for _, pv := range clusterPVs.Items {
```
Deleting a PV might take some time, it's not guaranteed to be done when `c.Core().PersistentVolumeClaims(ns).Delete(tmpClaim.Name)` returns. You need add another loop, trying several times with a sleep between the attempts, trying for several minutes:

```go
for i := 0; i < 30 && len(regressedPVs) > 0; i++ {
  By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
  ...
  time.Sleep(10)
}

copejon replied:
Got it.  I was toying with WaitForPersistentVolumePhase but your method is much better.

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			}()
+			// If a regression is detected, fail the test
+			if len(regressedPVs) > 0 {
+				framework.Failf("Found %d residual persistent volumes associated with StorageClass %s", len(regressedPVs), class.Name)
```
Please log all PV names so we can check the logs.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			regressedPVs := make([]v1.PersistentVolume, 0)
+			By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
+			// Every 10 seconds, check for PVs that have not been Released.  These indicate a regression
+			for i := 0; i < 30 && len(regressedPVs) == 0; i++ {
```
would it be better for 30 to be a `const`?

jsafrane replied:
`i < 30 && len(regressedPVs) == 0` looks suspicious. The loop should cycle while `len(regressedPVs) > 0` and stop when everything has been deleted (i.e. len==0)

copejon replied:
Okay, I see your intent now.  The issue with `for i := 0; i < 30 && len(regressedPVs) > 0; i++ {...}` is that `regressedPVs` starts at 0.  So the loop doesn't iterate.  I'll work out the loop to iterate until no PVs attached to the SC are detected.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			regressedPVs := make([]v1.PersistentVolume, 0)
```
delete -- see comment below

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
+			// Every 10 seconds, check for PVs that have not been Released.  These indicate a regression
+			for i := 0; i < 30 && len(regressedPVs) == 0; i++ {
+				time.Sleep(10 * time.Second)
```
should the polling wait also be a `const`?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+						pv.Status.Phase != v1.VolumeReleased {
+						// Any PVs associated with this storage class that are not Released
+						// indicate a regression
+						regressedPVs = append(regressedPVs, pv)
```
`regressedPVs = append(regressedPVs, pv.Name)`

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			defer func() {
+				for _, pv := range regressedPVs {
+					deletePDWithRetry(pv.Spec.PersistentVolumeSource.GCEPersistentDisk.PDName)
+					deletePersistentVolume(c, pv.Name)
```
`pvName`

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				}
+			}()
+			// If a regression is detected, fail the test
+			if regressionCount := len(regressedPVs); regressionCount > 0 {
```
`if len(regressedPVs) > 0 {...`

copejon replied:
Okay.  I was thinking it's more efficient to capture the integer from `len` than to call `len` 2 more times in the block below.

jeffvance replied:
True. THis is a left-over comment from when I was thinking out loud about a bigger refactoring of the `for` loop, and then deleted all of those comments but forgot this one...

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -131,6 +132,67 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				Expect(err).NotTo(HaveOccurred())
+				for _, pv := range clusterPVs.Items {
+					if pv.Annotations[storageutil.StorageClassAnnotation] == class.Name &&
+						pv.Status.Phase != v1.VolumeReleased {
```
I thought Jan's comment was to not bother testing Phase. Just `List` all pvs matching the sc annotation and expect the count to be zero eventually.

jsafrane replied:
Yes, no check for `Status`. Any PV with any status is bad.

And if I look correctly, the test waits for 300 seconds, which is IMO fine.

copejon replied:
I see. Will correct it.

copejon commented:
Jenkins GCI GCE e2e and Jenkins Kubermark GCE e2e need to be rerun.  Failures are in cluster setup, not test related.

copejon commented:
@jsafrane ptal

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -187,62 +189,66 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
-			if regressionCount := len(regressedPVs); regressionCount > 0 {
-				framework.Logf("Found %d residual persistent volumes associated with StorageClass %s:", regressionCount, class.Name)
+			// wait timeout err does not return a descriptive message.  Dress it up before failing.
+			if err != nil || regressCount > 0 {
```
please log the error, just in case it's not an timeout.

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -187,62 +189,66 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 
-			var raceAttempts int = 10
-			var suffix string = ns[len(ns)-5:]
+			const RaceAttempts int = 100
```
why capital 'R'? lowercase would be more appropriate here.

copejon replied:
Okay, I had thought that was a k8s standard for constants.  Will fix

jsafrane commented:
please fix two small nits and squash the commits.

jsafrane commented:
@k8s-bot test this
"Jenkins overloaded. Please try again later" :-\

jsafrane commented:
/lgtm

jeffvance commented:
@k8s-bot test this

copejon commented:
@k8s-bot test this

copejon commented:
@k8s-bot bazel test this

krousey commented:
/approve for test

copejon commented:
Let's give this a go
/assign @jeffvance

jeffvance commented:
@copejon random question: on L166, `if unmanagedZone, popped = unmanagedZones.PopAny(); !popped {` , you skip the test if there are no unmanaged zones. What is the probability of having no unmanaged zones? Is there a way to create an unmanaged zone in the test to guarantee that it runs?

jeffvance commented:
L181: `defer Expect(c.Core().PersistentVolumeClaims(ns).Delete(pvc.Name, nil)).To(Succeed())` Isn't this the same `defer` problem seen before where the args to `Expect` are evaluated when the `defer` is first processed? Thus, the pvc will be deleted.

jeffvance commented:
L178: `By("Creating a claim and expecting it to timeout")`
I see the test is labeled "[Slow]", but is there a way to explicitly test for failure rather than waiting? Probably not but just checking...

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 			framework.Logf(err.Error())
 		})
+
+		It("should test that deleting a claim before the volume is provisioned deletes the volume. [Volume]", func() {
```
should this be "[Slow]"?

copejon replied:
Barring regression, this test completes in under 5 minutes.  So according to the description for the tag, I think it can be omitted.

jeffvance replied:
Ok, I was wondering because of creating 30 claims and their provisioned pvs, and then deleting them all with a wait for each delete...

jeffvance replied:
L204-205: you want an `Excpect` for the `StorageClasses.Create` call and for the `StorageClasses().Delete` call. The `defer` should probably be a func() {...}

copejon commented:
@jeffvance I don't think you'd want to (or could) create a zone.  That would entail calling to the GCE api to alter the project, which seems unsafe.  Instead, the test looks for any zone which exists but is not tracked by k8s.  An alternative, perhaps less reliable method, would be to make up a zone name.  However, this could create false negatives in testing.  (regression occures but is missed b/c the pv fails to create in fake zone)

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			// NOTE:  Polls until no PVs are detected, times out at 5 minutes.
+
+			const raceAttempts int = 100
+			var suffix string = ns[len(ns)-5:] // match trailing random string of generated namespace
```
why 5? What if `ns` is shorter than 5 (unlikely)? Why not a random number between X and `len(ns)`?

copejon replied:
That works equally well.  In either case, the test just needs a randomized string to make the SC identifiable.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			var regressedPVs []v1.PersistentVolume
+			var regressCount int
+
+			By("Creating a StorageClass")
```
Can the code to create and verify a SC be move to its own func and shared by the `It("should not provision a volume in an unmanaged GCE zone` above too?

copejon replied:
Essentially it is.  The SC api object is supplied by `newStorageClass(...)`.  Another func to wrap create & expect only saves a couple lines of code.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			class := newStorageClass("kubernetes.io/gce-pd", suffix)
+			class, err := c.Storage().StorageClasses().Create(class)
+			Expect(err).NotTo(HaveOccurred())
+			defer c.Storage().StorageClasses().Delete(class.Name, nil)
```
Is this better?
```
defer func() {
         err = c.Storage().StorageClasses().Delete(class.Name, nil)
         Expect(err).ToBe(Nil))
} 
```

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			By(fmt.Sprintf("Creating and deleting PersistentVolumeClaims %d times", raceAttempts))
+			claim := newClaim(ns, suffix, false)
+			for i := 0; i < raceAttempts; i++ {
+				tmpClaim, err := c.Core().PersistentVolumeClaims(ns).Create(claim)
```
why not call `framework.CreatePVC`?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			for i := 0; i < raceAttempts; i++ {
+				tmpClaim, err := c.Core().PersistentVolumeClaims(ns).Create(claim)
+				Expect(err).NotTo(HaveOccurred())
+				err = c.Core().PersistentVolumeClaims(ns).Delete(tmpClaim.Name, nil)
```
and `framework.DeletePersistentVolumeClaim`?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+
+			By(fmt.Sprintf("Checking for residual PersistentVolumes associated with StorageClass %s", class.Name))
+			// Every 10 seconds, check for PVs that have not been deleted.
+			err = wait.Poll(10*time.Second, 300*time.Second, func() (bool, error) {
```
maybe a `const` for the total wait time?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+
+			const raceAttempts int = 100
+			var suffix string = ns[len(ns)-5:] // match trailing random string of generated namespace
+			var regressedPVs []v1.PersistentVolume
```
slice of pvs or is it better to have a slice of *pv?

copejon replied:
Either way.  `List()` returns a a ptr to a List object that contains a slice of PV objects (not ptrs).  So I can append them directly or reference them.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			// Every 10 seconds, check for PVs that have not been deleted.
+			err = wait.Poll(10*time.Second, 300*time.Second, func() (bool, error) {
+				regressedPVs = []v1.PersistentVolume{}
+				regressCount = 0
```
don't think this is needed. Can just check `len(regressedPVs)`

copejon replied:
Can you explain a bit? Do you mean `regressCount = len(regressedPVs)`?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				regressCount = 0
+
+				allPVs, err := c.Core().PersistentVolumes().List(metav1.ListOptions{})
+				Expect(err).NotTo(HaveOccurred())
```
Not sure you want an `Expect` in this poll func. I think it might be better to return `err` which will stop the poll loop.

copejon replied:
That's good point.  Will move it out.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			// REGRESSION: Deleting the PVC before the PV is provisioned can result in the PV
+			// not being deleted.
+			// NOTE:  Polls until no PVs are detected, times out at 5 minutes.
+
```
do you want a `framework.SkipUnlessProviderIs("gce", "gke")` here, like in the above `It` 's? Maybe this `SkipUnless...` should be in a `BeforeEach` somewhere?
`

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			}()
+			if regressCount > 0 {
+				framework.Logf("Remaining PersistentVolumes:")
+				for i, pv := range regressedPVs {
```
could just `Logf` the entire slice rather than each individual pv

copejon replied:
true but it prints it all on one line, which can be difficult to parse with a large-ish number of pvs.  It may not really matter though.

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -184,6 +186,70 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+				}
+				framework.Failf("Expected 0 PersistentVolumes remaining. Found %d", regressCount)
+			}
+			Expect(err).NotTo(HaveOccurred()) // wait.Poll returned error but no PVs remain
```
why the `Expect` here?

jeffvance commented:
/approve

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -218,6 +220,51 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			framework.SkipUnlessProviderIs("gce", "gke")
+
+			const raceAttempts int = 100
+			var suffix string = ns
```
maybe a comment saying you want a random string here?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -513,7 +512,7 @@ func newStorageClass(pluginName, suffix string) *storage.StorageClass {
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name: "myclass-" + suffix,
+			GenerateName: "sc-" + suffix + "-",
```
`suffix` is sometimes passed in empty. Is it ok for the generated name to begin with "sc--"?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -304,12 +303,12 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 			By("creating an external dynamic provisioner pod")
 			pod := startExternalProvisioner(c, ns)
-			defer c.Core().Pods(ns).Delete(pod.Name, nil)
+			defer framework.DeletePodOrFail(c, ns, pod.Name)
```
Would be nice to create a separate pr that allows for `apierrs.IsNotFound` in `DeletePodOrFail`.

copejon replied:
agreed. I'll ping you with the PR when it's up

jeffvance commented:
squash and it lgtm.

copejon committed:
Volume Provisioning E2E: test PVC delete causes PV delete

Removed wait for PVC phase Pending.

iterate test 100 times to increase chance of regression

Moved claim obj assignment out of loop.

add wait loop check for PVs

loop until no PVs detected

refactor per git comments

replace api calls with framework wrappers

add default suffix

copejon commented:
@k8s-bot unit test this

jeffvance commented:
/lgtm
